src/libs/elektra/hooks.c

```diff
@@ -101,26 +101,68 @@ static int initHooksSpec (KDB * kdb, Plugin * plugin, Key * errorKey)
 	return 0;
 }
 
+static KeySet * getSendNotificationHooksEnforcedByContract (const KeySet * contract)
+{
+	KeySet * returned = ksNew (0, KS_END);
+
+	KeySet * dupContract = ksDup (contract);
+
+	if (ksLookupByName (dupContract, "system:/elektra/contract/mountglobal/internalnotification", 0) != NULL)
+	{
+		ksAppendKey (returned, keyNew ("system:/elektra/hook/notification/send/plugins/byContractInternalnotification", KEY_VALUE,
+					       "internalnotification", KEY_END));
+	}
+
+	ksDel (dupContract);
+
+	return returned;
+}
+
 static int initHooksSendNotifications (KDB * kdb, const KeySet * config, KeySet * modules, const KeySet * contract, Key * errorKey)
 {
 	SendNotificationHook * lastHook = kdb->hooks.sendNotification;
 
+	KeySet * pluginsToLoad = getSendNotificationHooksEnforcedByContract (contract);
+	ksAppend (pluginsToLoad, config);
+
 	Key * pluginsKey = keyNew ("system:/elektra/hook/notification/send/plugins", KEY_END);
 
 	// Iterate through the config and find all direct children below system:/elektra/hook/notification/send/plugins
 	// This is defined as an array, so the elements should be like
 	//   - system:/elektra/hook/notification/send/plugins/#0
 	//   - system:/elektra/hook/notification/send/plugins/#1
 	// We actually don't check for the # to be present, so it will currently work on all keys directly below.
-	for (elektraCursor end, it = ksFindHierarchy (config, pluginsKey, &end); it < end; it++)
+	for (elektraCursor end, it = ksFindHierarchy (pluginsToLoad, pluginsKey, &end); it < end; it++)
 	{
-		Key * cur = ksAtCursor (config, it);
+		Key * cur = ksAtCursor (pluginsToLoad, it);
 		if (keyIsDirectlyBelow (pluginsKey, cur) == 0)
 		{
 			continue;
 		}
 
 		const char * pluginName = keyString (cur);
+
+		// Check whether this plugin has already been loaded
+		// TODO: This makes this function have a runtime complexity of O(n^2).
```

Comment:
Should be fine, there will never be so many notification plugins that this really matters.

Reply:
Yes, just leaving it in so that if someone sometimes decides to look for some optimizations, there is some work to do ;)